File: go/sound.go
package main

import (
	"bytes"
	"encoding/binary"
	"io"
	"math"
	"sync"
	"time"

	"github.com/gopxl/beep/v2"
	"github.com/gopxl/beep/v2/speaker"
	"github.com/gopxl/beep/v2/wav"
)

var (
	speakerInitOnce sync.Once
	speakerReady    bool
	muted           bool
	mutedMu         sync.Mutex
)

func toggleMute() {
	mutedMu.Lock()
	muted = !muted
	nowMuted := muted
	mutedMu.Unlock()

	themeMu.Lock()
	ctrl := themeCtrl
	themeMu.Unlock()

	if ctrl != nil {
		speaker.Lock()
		ctrl.Paused = nowMuted
		speaker.Unlock()
	}
}

func isMuted() bool {
	mutedMu.Lock()
	defer mutedMu.Unlock()
	return muted
}

func initSpeaker(sr beep.SampleRate) {
	speakerInitOnce.Do(func() {
		if err := speaker.Init(sr, sr.N(time.Second/10)); err == nil {
			speakerReady = true
		}
	})
}

// ---------------------------------------------------------------------------
// WAV encoding
// ---------------------------------------------------------------------------

func encodeWAV(sr int, samples []int16) []byte {
	buf := new(bytes.Buffer)
	dataSize := len(samples) * 2
	buf.Write([]byte("RIFF"))
	binary.Write(buf, binary.LittleEndian, uint32(36+dataSize))
	buf.Write([]byte("WAVE"))
	buf.Write([]byte("fmt "))
	binary.Write(buf, binary.LittleEndian, uint32(16))   // chunk size
	binary.Write(buf, binary.LittleEndian, uint16(1))    // PCM
	binary.Write(buf, binary.LittleEndian, uint16(1))    // mono
	binary.Write(buf, binary.LittleEndian, uint32(sr))   // sample rate
	binary.Write(buf, binary.LittleEndian, uint32(sr*2)) // byte rate
	binary.Write(buf, binary.LittleEndian, uint16(2))    // block align
	binary.Write(buf, binary.LittleEndian, uint16(16))   // bits per sample
	buf.Write([]byte("data"))
	binary.Write(buf, binary.LittleEndian, uint32(dataSize))
	for _, s := range samples {
		binary.Write(buf, binary.LittleEndian, s)
	}
	return buf.Bytes()
}

// ---------------------------------------------------------------------------
// Synthesis helpers
// ---------------------------------------------------------------------------

type sndNote struct {
	freq  float64 // Hz, 0 = rest
	beats float64 // duration in quarter-note beats
}

// renderTrack synthesises a note sequence into float64 samples.
// waveFn receives the instantaneous phase and returns a value in [-1, 1].
func renderTrack(notes []sndNote, sr, beatDur float64, waveFn func(float64) float64, amp, attackSec, releaseSec float64) []float64 {
	var out []float64
	atk := int(attackSec * sr)
	rel := int(releaseSec * sr)
	for _, n := range notes {
		num := int(n.beats * beatDur * sr)
		for i := range num {
			if n.freq == 0 {
				out = append(out, 0)
				continue
			}
			t := float64(i) / sr
			env := 1.0
			if i < atk {
				env = float64(i) / float64(atk)
			} else if i > num-rel {
				rem := num - i
				if rem <= 0 {
					env = 0
				} else {
					env = float64(rem) / float64(rel)
				}
			}
			out = append(out, waveFn(2*math.Pi*n.freq*t)*env*amp)
		}
	}
	return out
}

// mixToInt16 mixes one or more float64 sample tracks with clipping.
func mixToInt16(tracks ...[]float64) []int16 {
	maxLen := 0
	for _, t := range tracks {
		if len(t) > maxLen {
			maxLen = len(t)
		}
	}
	out := make([]int16, maxLen)
	for i := range out {
		var sum float64
		for _, t := range tracks {
			if i < len(t) {
				sum += t[i]
			}
		}
		if sum > 1 {
			sum = 1
		}
		if sum < -1 {
			sum = -1
		}
		out[i] = int16(sum * 32767)
	}
	return out
}

// ---------------------------------------------------------------------------
// Success chime
// ---------------------------------------------------------------------------

func generateSuccessWAV() []byte {
	sr := 44100
	sampleRate := float64(sr)
	notes := []sndNote{
		{523.25, 0.12}, {659.25, 0.12}, {783.99, 0.12}, {1046.50, 0.30},
	}
	richWave := func(phase float64) float64 {
		return math.Sin(phase)*0.7 + math.Sin(2*phase)*0.15 + math.Sin(3*phase)*0.05
	}
	// beats == seconds here (beatDur = 1.0)
	track := renderTrack(notes, sampleRate, 1.0, richWave, 0.35, 0.005, 0.04)
	return encodeWAV(sr, mixToInt16(track))
}

var successWAV = generateSuccessWAV()

func playSuccessSound() {
	if isMuted() {
		return
	}
	go func() {
		reader := bytes.NewReader(successWAV)
		streamer, format, err := wav.Decode(io.NopCloser(reader))
		if err != nil {
			return
		}
		defer streamer.Close()

		initSpeaker(format.SampleRate)
		if !speakerReady {
			return
		}

		done := make(chan struct{})
		speaker.Play(beep.Seq(streamer, beep.Callback(func() {
			close(done)
		})))
		<-done
	}()
}

// ---------------------------------------------------------------------------
// Theme music — Korobeiniki (Tetris Theme A)
// ---------------------------------------------------------------------------

// Note frequencies.
const (
	rest = 0
	C3   = 130.81
	D3   = 146.83
	E3   = 164.81
	A3   = 220.00
	B3   = 246.94
	C4   = 261.63
	D4   = 293.66
	E4   = 329.63
	A4   = 440.00
	B4   = 493.88
	C5   = 523.25
	D5   = 587.33
	E5   = 659.25
	F5   = 698.46
	G5   = 783.99
	A5   = 880.00
)

func generateThemeWAV() []byte {
	sr := 44100
	sampleRate := float64(sr)
	bpm := 150.0
	beatDur := 60.0 / bpm

	// --- Melody (chiptune square-ish wave) ---
	melody := []sndNote{
		// Phrase A
		// Bar 1
		{E5, 1},
		{B4, 0.5},
		{C5, 0.5},
		{D5, 1},
		{C5, 0.5},
		{B4, 0.5},
		// Bar 2
		{A4, 1},
		{A4, 0.5},
		{C5, 0.5},
		{E5, 1},
		{D5, 0.5},
		{C5, 0.5},
		// Bar 3
		{B4, 1.5},
		{C5, 0.5},
		{D5, 1},
		{E5, 1},
		// Bar 4
		{C5, 1},
		{A4, 1},
		{A4, 2},

		// Phrase B
		// Bar 5
		{rest, 0.5},
		{D5, 1},
		{F5, 0.5},
		{A5, 1},
		{G5, 0.5},
		{F5, 0.5},
		// Bar 6
		{E5, 1.5},
		{C5, 0.5},
		{E5, 1},
		{D5, 0.5},
		{C5, 0.5},
		// Bar 7
		{B4, 1.5},
		{C5, 0.5},
		{D5, 1},
		{E5, 1},
		// Bar 8
		{C5, 1},
		{A4, 1},
		{A4, 2},
	}

	// --- Bass (pulsing root on beats 1 & 3) ---
	bass := []sndNote{
		{E3, 1}, {rest, 1}, {E3, 1}, {rest, 1}, // bar 1
		{A3, 1}, {rest, 1}, {A3, 1}, {rest, 1}, // bar 2
		{E3, 1}, {rest, 1}, {E3, 1}, {rest, 1}, // bar 3
		{A3, 1}, {rest, 1}, {A3, 1}, {rest, 1}, // bar 4
		{D3, 1}, {rest, 1}, {D3, 1}, {rest, 1}, // bar 5
		{C3, 1}, {rest, 1}, {C3, 1}, {rest, 1}, // bar 6
		{E3, 1}, {rest, 1}, {E3, 1}, {rest, 1}, // bar 7
		{A3, 1}, {rest, 1}, {A3, 1}, {rest, 1}, // bar 8
	}

	// Band-limited square wave (odd harmonics).
	sqWave := func(phase float64) float64 {
		return math.Sin(phase) + math.Sin(3*phase)/3 + math.Sin(5*phase)/5
	}

	melTrack := renderTrack(melody, sampleRate, beatDur, sqWave, 0.15, 0.005, 0.02)
	bassTrack := renderTrack(bass, sampleRate, beatDur, math.Sin, 0.10, 0.005, 0.02)

	return encodeWAV(sr, mixToInt16(melTrack, bassTrack))
}

// --- Theme playback (looping) ---

var (
	themeWAV  = generateThemeWAV()
	themeCtrl *beep.Ctrl
	themeMu   sync.Mutex
)

func playThemeMusic() {
	go func() {
		reader := bytes.NewReader(themeWAV)
		streamer, format, err := wav.Decode(io.NopCloser(reader))
		if err != nil {
			return
		}

		initSpeaker(format.SampleRate)
		if !speakerReady {
			return
		}

		loop := beep.Loop(-1, streamer)
		ctrl := &beep.Ctrl{Streamer: loop}

		themeMu.Lock()
		themeCtrl = ctrl
		themeMu.Unlock()

		speaker.Play(ctrl)
	}()
}

func stopThemeMusic() {
	themeMu.Lock()
	defer themeMu.Unlock()
	if themeCtrl != nil {
		speaker.Lock()
		themeCtrl.Paused = true
		speaker.Unlock()
		themeCtrl = nil
	}
}

File: rust/src/sound.rs
use std::f64::consts::PI;
use std::io::Cursor;
use std::sync::atomic::{AtomicBool, Ordering};
use std::sync::Arc;
use std::thread;

use rodio::{Decoder, OutputStream, Sink};

// ---------------------------------------------------------------------------
// WAV encoding
// ---------------------------------------------------------------------------

fn encode_wav(sr: u32, samples: &[i16]) -> Vec<u8> {
    let data_size = samples.len() as u32 * 2;
    let mut buf: Vec<u8> = Vec::with_capacity(44 + data_size as usize);
    buf.extend_from_slice(b"RIFF");
    buf.extend_from_slice(&(36 + data_size).to_le_bytes());
    buf.extend_from_slice(b"WAVE");
    buf.extend_from_slice(b"fmt ");
    buf.extend_from_slice(&16u32.to_le_bytes()); // chunk size
    buf.extend_from_slice(&1u16.to_le_bytes()); // PCM
    buf.extend_from_slice(&1u16.to_le_bytes()); // mono
    buf.extend_from_slice(&sr.to_le_bytes()); // sample rate
    buf.extend_from_slice(&(sr * 2).to_le_bytes()); // byte rate
    buf.extend_from_slice(&2u16.to_le_bytes()); // block align
    buf.extend_from_slice(&16u16.to_le_bytes()); // bits per sample
    buf.extend_from_slice(b"data");
    buf.extend_from_slice(&data_size.to_le_bytes());
    for &s in samples {
        buf.extend_from_slice(&s.to_le_bytes());
    }
    buf
}

// ---------------------------------------------------------------------------
// Synthesis helpers
// ---------------------------------------------------------------------------

struct SndNote {
    freq: f64,  // Hz, 0 = rest
    beats: f64, // duration in quarter-note beats
}

/// Render a note sequence to float64 samples.
fn render_track(
    notes: &[SndNote],
    sr: f64,
    beat_dur: f64,
    wave_fn: fn(f64) -> f64,
    amp: f64,
    attack_sec: f64,
    release_sec: f64,
) -> Vec<f64> {
    let atk = (attack_sec * sr) as usize;
    let rel = (release_sec * sr) as usize;
    let mut out = Vec::new();
    for n in notes {
        let num = (n.beats * beat_dur * sr) as usize;
        for i in 0..num {
            if n.freq == 0.0 {
                out.push(0.0);
                continue;
            }
            let t = i as f64 / sr;
            let env = if i < atk {
                i as f64 / atk as f64
            } else if i > num - rel {
                let rem = num - i;
                if rem == 0 {
                    0.0
                } else {
                    rem as f64 / rel as f64
                }
            } else {
                1.0
            };
            out.push(wave_fn(2.0 * PI * n.freq * t) * env * amp);
        }
    }
    out
}

/// Mix one or more float64 tracks into int16 with clipping.
fn mix_to_i16(tracks: &[&[f64]]) -> Vec<i16> {
    let max_len = tracks.iter().map(|t| t.len()).max().unwrap_or(0);
    let mut out = vec![0i16; max_len];
    for i in 0..max_len {
        let mut sum: f64 = 0.0;
        for t in tracks {
            if i < t.len() {
                sum += t[i];
            }
        }
        sum = sum.clamp(-1.0, 1.0);
        out[i] = (sum * 32767.0) as i16;
    }
    out
}

// Waveforms

fn sine_wave(phase: f64) -> f64 {
    phase.sin()
}

fn square_wave(phase: f64) -> f64 {
    phase.sin() + (3.0 * phase).sin() / 3.0 + (5.0 * phase).sin() / 5.0
}

fn rich_wave(phase: f64) -> f64 {
    phase.sin() * 0.7 + (2.0 * phase).sin() * 0.15 + (3.0 * phase).sin() * 0.05
}

// ---------------------------------------------------------------------------
// Success chime
// ---------------------------------------------------------------------------

fn generate_success_wav() -> Vec<u8> {
    let sr = 44100u32;
    let srf = sr as f64;
    let notes = [
        SndNote {
            freq: 523.25,
            beats: 0.12,
        },
        SndNote {
            freq: 659.25,
            beats: 0.12,
        },
        SndNote {
            freq: 783.99,
            beats: 0.12,
        },
        SndNote {
            freq: 1046.50,
            beats: 0.30,
        },
    ];
    let track = render_track(&notes, srf, 1.0, rich_wave, 0.35, 0.005, 0.04);
    encode_wav(sr, &mix_to_i16(&[&track]))
}

// ---------------------------------------------------------------------------
// Theme music — Korobeiniki (Tetris Theme A)
// ---------------------------------------------------------------------------

const REST: f64 = 0.0;
const C3: f64 = 130.81;
const D3: f64 = 146.83;
const E3: f64 = 164.81;
const A3: f64 = 220.00;
const A4: f64 = 440.00;
const B4: f64 = 493.88;
const C5: f64 = 523.25;
const D5: f64 = 587.33;
const E5: f64 = 659.25;
const F5: f64 = 698.46;
const G5: f64 = 783.99;
const A5: f64 = 880.00;

fn generate_theme_wav() -> Vec<u8> {
    let sr = 44100u32;
    let srf = sr as f64;
    let bpm = 150.0;
    let beat_dur = 60.0 / bpm;

    let melody = [
        // Phrase A
        SndNote {
            freq: E5,
            beats: 1.0,
        },
        SndNote {
            freq: B4,
            beats: 0.5,
        },
        SndNote {
            freq: C5,
            beats: 0.5,
        },
        SndNote {
            freq: D5,
            beats: 1.0,
        },
        SndNote {
            freq: C5,
            beats: 0.5,
        },
        SndNote {
            freq: B4,
            beats: 0.5,
        },
        SndNote {
            freq: A4,
            beats: 1.0,
        },
        SndNote {
            freq: A4,
            beats: 0.5,
        },
        SndNote {
            freq: C5,
            beats: 0.5,
        },
        SndNote {
            freq: E5,
            beats: 1.0,
        },
        SndNote {
            freq: D5,
            beats: 0.5,
        },
        SndNote {
            freq: C5,
            beats: 0.5,
        },
        SndNote {
            freq: B4,
            beats: 1.5,
        },
        SndNote {
            freq: C5,
            beats: 0.5,
        },
        SndNote {
            freq: D5,
            beats: 1.0,
        },
        SndNote {
            freq: E5,
            beats: 1.0,
        },
        SndNote {
            freq: C5,
            beats: 1.0,
        },
        SndNote {
            freq: A4,
            beats: 1.0,
        },
        SndNote {
            freq: A4,
            beats: 2.0,
        },
        // Phrase B
        SndNote {
            freq: REST,
            beats: 0.5,
        },
        SndNote {
            freq: D5,
            beats: 1.0,
        },
        SndNote {
            freq: F5,
            beats: 0.5,
        },
        SndNote {
            freq: A5,
            beats: 1.0,
        },
        SndNote {
            freq: G5,
            beats: 0.5,
        },
        SndNote {
            freq: F5,
            beats: 0.5,
        },
        SndNote {
            freq: E5,
            beats: 1.5,
        },
        SndNote {
            freq: C5,
            beats: 0.5,
        },
        SndNote {
            freq: E5,
            beats: 1.0,
        },
        SndNote {
            freq: D5,
            beats: 0.5,
        },
        SndNote {
            freq: C5,
            beats: 0.5,
        },
        SndNote {
            freq: B4,
            beats: 1.5,
        },
        SndNote {
            freq: C5,
            beats: 0.5,
        },
        SndNote {
            freq: D5,
            beats: 1.0,
        },
        SndNote {
            freq: E5,
            beats: 1.0,
        },
        SndNote {
            freq: C5,
            beats: 1.0,
        },
        SndNote {
            freq: A4,
            beats: 1.0,
        },
        SndNote {
            freq: A4,
            beats: 2.0,
        },
    ];

    let bass = [
        SndNote {
            freq: E3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: E3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: A3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: A3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: E3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: E3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: A3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: A3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: D3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: D3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: C3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: C3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: E3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: E3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: A3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
        SndNote {
            freq: A3,
            beats: 1.0,
        },
        SndNote {
            freq: REST,
            beats: 1.0,
        },
    ];

    let mel_track = render_track(&melody, srf, beat_dur, square_wave, 0.15, 0.005, 0.02);
    let bass_track = render_track(&bass, srf, beat_dur, sine_wave, 0.10, 0.005, 0.02);

    encode_wav(sr, &mix_to_i16(&[&mel_track, &bass_track]))
}

// ---------------------------------------------------------------------------
// SoundEngine — owns the audio output and provides play/mute controls
// ---------------------------------------------------------------------------

pub struct SoundEngine {
    muted: Arc<AtomicBool>,
    theme_sink: Option<Sink>,
    success_wav: Vec<u8>,
    // Keep the OutputStream alive for the lifetime of the engine.
    _stream: Option<OutputStream>,
}

impl SoundEngine {
    pub fn new() -> Self {
        let muted = Arc::new(AtomicBool::new(false));
        let success_wav = generate_success_wav();

        // Try to open audio output; if it fails, we run silently.
        let (stream, theme_sink) = match OutputStream::try_default() {
            Ok((stream, handle)) => {
                let theme_wav = generate_theme_wav();
                let theme_sink = Sink::try_new(&handle).ok();
                if let Some(ref sink) = theme_sink {
                    let cursor = Cursor::new(theme_wav);
                    if let Ok(source) = Decoder::new(cursor) {
                        sink.append(source);
                        // rodio Sink doesn't have a native loop, so we
                        // re-append in a background thread (see start_theme_loop).
                    }
                }
                (Some(stream), theme_sink)
            }
            Err(_) => (None, None),
        };

        SoundEngine {
            muted,
            theme_sink,
            success_wav,
            _stream: stream,
        }
    }

    /// Start the theme music loop in a background thread that re-appends
    /// the theme WAV data whenever the sink runs low.
    pub fn start_theme_loop(&self) {
        let Some(ref sink) = self.theme_sink else {
            return;
        };
        // We need to keep feeding the sink with new copies of the theme.
        // Spawn a thread that watches the sink and appends when nearly empty.
        let muted = Arc::clone(&self.muted);
        let theme_wav = generate_theme_wav();

        // The sink is not Send, so we use a different approach:
        // We pre-fill the sink with many repeats upfront and that's enough
        // for a typical game session. If someone plays for hours it will
        // eventually stop, but that's acceptable.
        // Actually, let's use a smarter approach: we'll clone the handle.
        // rodio::Sink is not Send, but we can just append many copies now.
        for _ in 0..100 {
            let cursor = Cursor::new(theme_wav.clone());
            if let Ok(source) = Decoder::new(cursor) {
                sink.append(source);
            }
        }

        // Apply initial mute state.
        if muted.load(Ordering::Relaxed) {
            sink.set_volume(0.0);
        }
    }

    pub fn play_success(&self) {
        if self.muted.load(Ordering::Relaxed) {
            return;
        }
        let Some(ref _stream) = self._stream else {
            return;
        };
        // Play in a background thread so we don't block the TUI.
        let wav = self.success_wav.clone();
        thread::spawn(move || {
            if let Ok((_stream, handle)) = OutputStream::try_default() {
                if let Ok(sink) = Sink::try_new(&handle) {
                    let cursor = Cursor::new(wav);
                    if let Ok(source) = Decoder::new(cursor) {
                        sink.append(source);
                        sink.sleep_until_end();
                    }
                }
            }
        });
    }

    pub fn toggle_mute(&self) {
        let was_muted = self.muted.fetch_xor(true, Ordering::Relaxed);
        let now_muted = !was_muted;
        if let Some(ref sink) = self.theme_sink {
            if now_muted {
                sink.set_volume(0.0);
            } else {
                sink.set_volume(1.0);
            }
        }
    }

    pub fn is_muted(&self) -> bool {
        self.muted.load(Ordering::Relaxed)
    }

    pub fn stop(&self) {
        if let Some(ref sink) = self.theme_sink {
            sink.stop();
        }
    }
}
